### ISSUE TITLE
VC4: Possible vertex shader problem

### ISSUE THREAD

**Issue by vanfanel** [anholt/mesa#110]
Hi Eric,

I have found that SDL2 (latest stable 2.0.8) shows a strange problem on the Pi: the image in any 2D game or program is rotated by a pixel. Now, it's an SDL2 problem of course, but it only happens on the Pi platform using the KMS/DRM + GLES2 backend, on the VC4 driver.
I have investigated the problem and it comes down to this simple vertex shader that SDL2 uses by default to draw any 2D image scaled using a GLES2 texture. It's in src/render/opengles2/SDL_shaders_gles2.c:

```
static const Uint8 GLES2_VertexSrc_Default_[] = " \ 
    uniform mat4 u_projection; \
    attribute vec2 a_position; \
    attribute vec2 a_texCoord; \
    attribute float a_angle; \
    attribute vec2 a_center; \
    varying vec2 v_texCoord; \
    \   
    void main() \
    { \ 
        float angle = radians(a_angle); \
        float c = cos(angle); \
        float s = sin(angle); \
        mat2 rotationMatrix = mat2(c, -s, s, c); \
        vec2 position = rotationMatrix * (a_position - a_center) + a_center; \
        v_texCoord = a_texCoord; \
        gl_Position = u_projection * vec4(position, 0.0, 1.0);\
        gl_PointSize = 1.0; \
    } \ 
";

```

Changing it so it does not premultiply `position` by `rotationMatrix`, the problem goes away and the image is not rotated anymore.
This is only happening on the Pi as I said, using the latest `mesa-18.1.6` sources, so that's why I believe it's related to the VC4 driver.
What do you think?

**Comment by schnitzeltony** [comment]
I'd love to see this solved - see https://github.com/anholt/mesa/issues/58

**Comment by anholt** [comment]
Wow, passing in the angle and computing the matrix in the shader instead of just passing in the matrix?  That's... a choice you could make.  But it at least gives an explanation for the problem that I could never figure out in #58.  Our sin approximation is probably off at 0, so with no rotation you get a bit anyway.  IRC recommended a few adjustments to our approximations, I'll have to try one of those out at some point.

(still mostly on leave currently)

**Comment by schnitzeltony** [comment]
I think passing in matrix could also be a performance win - what do you think? Will spend some time on that and see how far I get.

**Comment by schnitzeltony** [comment]
Sorry - for the noise caused by referencing this issue and doing multiple rebase.

I sent out a patch to SDL that moves sin/cos out of shader [1] and tested that for both: simple draw and SDL_RenderCopyEx (which is the only setter of angle - all others don't use angle).

So after VERY long time I am happy and consider this fixed - hope SDL people won't tear the patch to pieces...

Thanks everybody for taking time on this - I'll spend my evening with some retro games displayed undistorted :)

[1] https://bugzilla.libsdl.org/show_bug.cgi?id=4239

**Comment by vanfanel** [comment]
@schnitzeltony : I am happy we finally found out what the problem was, and I hope the SDL2 people includes accepts your patch ASAP. I don't understand how this has been ignored for so long... what's people doing with their Pi's?? SDL2 on the Pi has been running on the open source driver for years now....

**Comment by schnitzeltony** [comment]
@vanfanel I think it was not ignored - it was not easy to catch. You introduced the missing hint!

**Comment by anholt** [comment]
Pushed a fix, I think:

```
commit 10d5d2d527dea11f4afe300eebeaba077f169af0
Author: Eric Anholt <eric@anholt.net>
Date:   Fri Sep 21 14:11:12 2018 -0700

    vc4: Fix sin(0.0) and cos(0.0) accuracy to fix SDL rendering rotation.
```

The sdl2box app in #58 doesn't call sin or cos for me, so I didn't have an SDL test to use.  Feel free to reopen if this doesn't fix it!
